google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectLoader.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.project;
+
+import com.google.api.client.repackaged.com.google.common.base.Strings;
+import com.google.api.services.cloudresourcemanager.CloudResourceManager;
+import com.google.api.services.cloudresourcemanager.model.ListProjectsResponse;
+import com.google.api.services.cloudresourcemanager.model.Project;
+import com.google.cloud.tools.intellij.login.CredentialedUser;
+import com.google.cloud.tools.intellij.resources.GoogleApiClientFactory;
+import com.google.cloud.tools.intellij.util.GctBundle;
+import com.google.cloud.tools.intellij.util.ThreadUtil;
+import com.google.common.annotations.VisibleForTesting;
+import com.intellij.openapi.diagnostic.Logger;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+
+/** Loads list of {@link Project} for a {@link CredentialedUser}. */
+class ProjectLoader {
+  private static final Logger LOG = Logger.getInstance(ProjectLoader.class);
+
+  private static final int PROJECTS_MAX_PAGE_SIZE = 300;
+  private static final String PROJECT_DELETE_REQUESTED = "DELETE_REQUESTED";
+
+  /** Callback interface to receive project list results from async execution. */
+  interface ProjectLoaderResultCallback {
+    void projectListReady(List<Project> result);
+    void onError(String errorDetails);
+  }
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  void loadUserProjectsInBackground(CredentialedUser user, ProjectLoaderResultCallback resultCallback) {
+    ThreadUtil.getInstance()
+        .executeInBackground(
+            () -> loadUserProjects(user, resultCallback));
+  }
+
+  @VisibleForTesting
+  void loadUserProjects(CredentialedUser user, ProjectLoaderResultCallback resultCallback) {
```

Comment:
you can keep this private and test the async method above instead: unit tests that use @CloudToolsRule replace the executor used in ThreadUtil with a direct executor service.